### ISSUE TITLE
Memory limit

### ISSUE THREAD

**Issue by royfrancis** [hartleys/JunctionSeq#10]
Hi,
I've tried running JuntionSeq in R on my windows desktop computer (8 cores, 32 gb ram) and it took several days to run since it could use only one core. But atleast, it completed. Now I am trying to run it on a linux cluster. And it keeps failing due to overdraft on memory. I initially provided 8 cores and 32gb ram and it failed. Then I provided 12 cores and 48gb ram and it still fails.

```
---> STARTING estimateJunctionSeqDispersions: (v1.2.4) (Thu Jun 23 11:33:57 2016)
-----> ejsd: 527130 counting bins are marked 'testable'. across 25554 genes.
             (304333 exonic regions, 222797 known junctions, 0 novel junctions)
---------> Executing DESeq2 call: estimateUnsharedDispersions
slurmstepd: Job 10321260 exceeded memory limit (43966912 > 43008000), being killed
slurmstepd: Exceeded job memory limit
```

Is there some way to set a limit on memory? Or is this some completely unrelated issue?

**Comment by hartleys** [comment]
Can you post the exact command used?

Note that the built-in R parallel processing has a few implicit weaknesses that are basically unavoidable. One of the biggest weaknesses is that increasing the number of cores vastly increases the memory requirements, because R will create separate R processes for each core.

Other questions:
- What is your sample size?
- Species? 
- Average total read count? 
- How many replicates of each biological conditions do you have?

**Comment by royfrancis** [comment]
``` R
library(JunctionSeq)
library(BiocParallel)

ncores <- 12

ja <- readJunctionSeqCounts(countfiles = cpath,
                               samplenames = names,
                               design=data.frame(condition=treatment,tissue=tissue),
                               flat.gff.file = "dr-flat-stranded.gff",
                               use.novel.junctions=F,
                               analysis.type = "junctionsAndExons")
ja <- estimateJunctionSeqSizeFactors(ja)
# line that produces the error
ja <- estimateJunctionSeqDispersions(ja, nCores = ncores)
```

I have 52 samples. Read depth varies from 20M to 80M reads. My model has tissue and condition. I have 2 tissues and 2 conditions. So, 26 files are tissue a and 26 files tissue b. Within the 26, 13 are control and 13 are treatment. Organism is Zebrafish.

Now that you mentioned this, I am thinking I should perhaps run this on a smaller dataset first. I jumped into it because it ran without errors on DEXSeq. I could also split everything by tissue and run separately. I am interested in comparing control and treatment within tissue a and within tissue b. It's probably not that interesting to compare between tissues.

**Comment by hartleys** [comment]
JunctionSeq generally has higher resource requirements than DEXSeq, since it includes more data.

I would recommend splitting up the analyses between treatments. For one thing, in the analysis listed above, you haven't modified the [test.formula](http://hartleys.github.io/JunctionSeq/Rhtml/runJunctionSeqAnalyses.html) variables, so it's actually ignoring the tissue variable and comparing treatments with all tissues merged. 

If you want to analyze multiple tissues simultaneously you at least need to add it as a confounding variable. But really, I would not expect the treatment to affect the same genes in the same way across all tissues, so that wouldn't necessarily give you what you're actually looking for.

Cross-tissue comparisons are also difficult, because they often have very different expression profiles. Basically all expression methods (DESeq, DEXSeq, edgeR, CuffDiff, etc) make some sort of assumption that "most genes are not DE", which is vitally important for normalization. This assumption is often violated in cross-tissue analyses. Simply including different tissues in the size factor calculation step can skew your results.

Really you should run separate analyses comparing treatment in tissue A and then treatment in tissue B. You can then compare the results of these analyses if you like.

So, to summarize, my recommendations are:
- Run separate analyses for each tissue type.
- Reduce the number of cores used. This will _greatly_ reduce the memory usage.

Given the large size of your dataset, you might even want to split your data into a discovery and replication set. This would give you a solid confirmation, but would have less overall power. So it depends on what your resources and priorities are.

**Comment by royfrancis** [comment]
Thank you for your suggestions. I will give it a go with reduced dataset first.

**Comment by aiminy** [comment]
I am using runJunctionSeqAnalyses function, but I got the following error liking this:
Error in fitDisp(ySEXP = ySEXP, xSEXP = xSEXP, mu_hatSEXP = mu_hatSEXP,  : 
  not compatible with requested type

Do you any idea about this?

Aimin
